nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+//! use mm(or MM) to represent martrix market
+use crate::coo::CooMatrix;
+use crate::SparseFormatError;
+use crate::SparseFormatErrorKind;
+use nalgebra::Complex;
+use pest::iterators::Pairs;
+use pest::Parser;
+use std::cmp::PartialEq;
+use std::fmt;
+use std::fmt::Formatter;
+use std::fs;
+use std::num::ParseIntError;
+use std::ops::Neg;
+use std::path::Path;
+use std::str::FromStr;
+
+/// A description of the error that occurred during importing a matrix from a matrix market format data.
+#[derive(Debug)]
+pub struct MMError {
+    error_kind: MMErrorKind,
+    message: String,
+}
+
+/// Errors produced by functions that expect well-formed matrix market format data.
+/// > _NOTE1:_ Since the matrix market design didn't mention if duplicate entries are allowed or not, so, it's allowed here.
+///
+/// > _NOTE2:_ Dense matrices are not supported here. For example, `%%matrixmarket matrix array real general` will give `MMErrorKind::ParsingError`.
```

Comment:
It's just because I think reading dense matrix into a `nalgebra::Matrix` could be better? but if want to be read as a dense matrix, it can also be done.

Reply:
That's true - but it might be that the user doesn't know up front that the matrix market matrix is in fact stored as a dense matrix. For example, imagine that you have a large collection of MM files and you want to test some algorithm on it, then maybe some of those files have dense matrices. It would be nice if it "just works" in basically all cases, I think!